### ISSUE TITLE
waitpid() example in perlfunc(1) is bogus

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9250]
Migrated from [rt.perl.org#51642](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=51642) (status was 'resolved')

Searchable as RT51642$

**Comment by p5pRT** [comment]
### From @JulNadeauCA

#### Created by @JulNadeauCA


This is a bug report for perl from vedge@&#8203;hypertriton\.com\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
perlfunc\(1\) gives the following example for waitpid\(\) usage&#8203;:

&nbsp;    use POSIX "&#8203;:sys\_wait\_h";
&nbsp;    \#\.\.\.
&nbsp;    do \{
&nbsp;        $kid = waitpid\(\-1\, WNOHANG\);
&nbsp;    \} while $kid > 0;

This would obviously cause an infinite loop as waitpid\(\) returns
\-1 only when there are no more child processes to wait on\. The
correct usage would be&#8203;:

&nbsp;    use POSIX "&#8203;:sys\_wait\_h";
&nbsp;    \#\.\.\.
&nbsp;    do \{
&nbsp;        $kid = waitpid\(\-1\, WNOHANG\);
&nbsp;    \} while $kid \!= \-1;

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=medium

Site configuration information for perl v5.8.8:

Configured by root at Thu Jan  1  0:00:00 UTC 1970.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=openbsd, osvers=4.2, archname=amd64-openbsd
    uname='openbsd'
    config_args='-dsE -Dopenbsd_distribution=defined'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=define uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -fno-delete-null-pointer-checks -pipe -I/usr/local/include',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -fno-delete-null-pointer-checks -pipe -I/usr/local/include'
    ccversion='', gccversion='3.3.5 (propolice)', gccosandvers='openbsd4.2'
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-E '
    libpth=/usr/lib
    libs=-lm -lutil -lc
    perllibs=-lm -lutil -lc
    libc=/usr/lib/libc.a, so=so, useshrplib=true, libperl=libperl.so.10.1
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-R/usr/libdata/perl5/amd64-openbsd/5.8.8/CORE'
    cccdlflags='-DPIC -fPIC ', lddlflags='-shared -fPIC '

Locally applied patches:
    


@INC for perl v5.8.8:
    /usr/libdata/perl5/amd64-openbsd/5.8.8
    /usr/local/libdata/perl5/amd64-openbsd/5.8.8
    /usr/libdata/perl5
    /usr/local/libdata/perl5
    /usr/local/libdata/perl5/site_perl/amd64-openbsd
    /usr/libdata/perl5/site_perl/amd64-openbsd
    /usr/local/libdata/perl5/site_perl
    /usr/libdata/perl5/site_perl
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/vedge
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/vedge/bin:/bin:/usr/bin:/usr/local/bin:/usr/X11R6/bin:/sbin:/usr/games:/usr/sbin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @abigail

On Wed\, Mar 12\, 2008 at 03&#8203;:49&#8203;:25AM \-0700\, vedge @&#8203; hypertriton\. com wrote&#8203;:
> \# New Ticket Created by  vedge@&#8203;hypertriton\.com 
> \# Please include the string&#8203;:  \[perl \#51642\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=51642 >
> 
> 
> 
> This is a bug report for perl from vedge@&#8203;hypertriton\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> perlfunc\(1\) gives the following example for waitpid\(\) usage&#8203;:
> 
>     use POSIX "&#8203;:sys\_wait\_h";
>     \#\.\.\.
>     do \{
>         $kid = waitpid\(\-1\, WNOHANG\);
>     \} while $kid > 0;
> 
> This would obviously cause an infinite loop as waitpid\(\) returns
> \-1 only when there are no more child processes to wait on\. The
> correct usage would be&#8203;:
> 
>     use POSIX "&#8203;:sys\_wait\_h";
>     \#\.\.\.
>     do \{
>         $kid = waitpid\(\-1\, WNOHANG\);
>     \} while $kid \!= \-1;


The latter might actually result in an infinite loop\, as waitpid\(\) may
return 0\. 

But regardless of the actual return value of waitpid\(\)\, if the guard 
'$kid > 0' could cause an infinite loop\, so will the guard '$kid \!= \-1'\.
Your suggestion would cause the loop to terminate only if $kid == \-1\,
but if $kid == \-1\, the example will terminate as well\.


Abigail

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
